### ISSUE TITLE
mesolve giving erroneous results for time dependent Hamiltonian

### ISSUE THREAD

**Issue by nkvij** [qutip/qutip#1038]
I have been trying to simulate a 3 level system which has a time-dependent Hamiltonian. However, the plots I obtain are wrong. I am new to qutip so I might be missing something very silly.
Here's the code: 
```
from qutip import *
import numpy as np
import matplotlib.pyplot as plt

qd0 = basis(3,0)
qd1 = basis(3,1)
qd2 = basis(3,2)

hbar = 6.582*10**-4

times = np.linspace(0.0,1,100000)

sig00 = tensor(qd0*qd0.dag(), qeye(3))
sig11 = tensor(qd1*qd1.dag(), qeye(3))
sig22 = tensor(qd2*qd2.dag(), qeye(3))
sig01 = tensor(qd0*qd1.dag(), qeye(3))
sig02 = tensor(qd0*qd2.dag(), qeye(3))
sig12 = tensor(qd1*qd2.dag(), qeye(3))
sig21 = tensor(qd2*qd1.dag(), qeye(3))

a = tensor(qeye(3), destroy(3))

w1 = 2000.0/hbar
w2 = 2001.0/hbar
wph = 1.0/hbar
g = 120
pump1 = 10.0
pump2 = 1
gamma = 10**-3
nth = 10**-6
y = 0*times + nth

T1 = 0.001
T2 = 1

H01  = w1*sig11 + w2*sig22 + wph*a.dag()*a + g*(sig12*a.dag() + sig21*a)
H1a = pump1*sig01
H2a = pump1*sig01.dag()

H1b = pump2*sig01
H2b = pump2*sig01.dag()

def H1_coeff(times, args):
    return np.exp(w1*times*1j)
def H2_coeff(times, args):
    return np.exp(-w1*times*1j)

Ha = [H01, [H1a, H1_coeff],[H2a, H2_coeff]]
Hb = [H01, [H1b, H1_coeff],[H2b, H2_coeff]]

psi0 = tensor(qd0, fock(3,1))

result1 = mesolve(Ha, psi0, times, [np.sqrt(T1)*sig01, np.sqrt(T2)*sig02, np.sqrt(gamma*(nth + 1))*a, np.sqrt(gamma*nth)*a.dag()], [a.dag()*a, sig00, sig11, sig22], progress_bar=True)
result2 = mesolve(Hb, psi0, times, [np.sqrt(T1)*sig01, np.sqrt(T2)*sig02, np.sqrt(gamma*(nth + 1))*a, np.sqrt(gamma*nth)*a.dag()], [a.dag()*a, sig00, sig11, sig22], progress_bar=True)


plt.figure()
plt.plot(times, result1.expect[0], label = 'pump = 10')
plt.plot(times, result2.expect[0], label = 'pump = 1')
plt.plot(times, y,'k')
plt.xlabel('Time (ns)')
plt.ylabel('Phonon number')
legend = plt.legend()
plt.savefig("phonon number.png")
plt.savefig("phonon number.pdf")
output_data = np.vstack((times, result1.expect[0]))
file_data_store('phonon number.dat',output_data.T, numtype = "real")
output_data = np.vstack((times, result2.expect[0]))
file_data_store('phonon number.dat',output_data.T, numtype = "real")
plt.title('g=120, nth=$10^{-6}$, T1=$10^{-3}$, T2=1')

plt.figure()
plt.plot(times, result1.expect[1], label = 'pump = 10')
plt.plot(times, result2.expect[1], label = 'pump = 1')
plt.xlabel('Time (ns)')
plt.ylabel('$\sigma_{00}$')
legend = plt.legend()
plt.savefig("sigma_00.png")
plt.savefig("sigma_00.pdf")
output_data = np.vstack((times, result1.expect[1]))
file_data_store('sigma_00.dat',output_data.T, numtype = "real")
output_data = np.vstack((times, result2.expect[1]))
file_data_store('sigma_00.dat',output_data.T, numtype = "real")
plt.title('g=120, nth=$10^{-6}$, T1=$10^{-3}$, T2=1')


plt.figure()
plt.plot(times, result1.expect[2], label = 'pump = 10')
plt.plot(times, result2.expect[2], label = 'pump = 1')
plt.xlabel('Time (ns)')
plt.ylabel('$\sigma_{11}$')
legend = plt.legend()
plt.savefig("sigma_11.png")
plt.savefig("sigma_11.pdf")
output_data = np.vstack((times, result1.expect[2]))
file_data_store('sigma_11.dat',output_data.T, numtype = "real")
output_data = np.vstack((times, result2.expect[2]))
file_data_store('sigma_11.dat',output_data.T, numtype = "real")
plt.title('g=120, nth=$10^{-6}$, T1=$10^{-3}$, T2=1')


plt.figure()
plt.plot(times, result1.expect[3], label = 'pump = 10')
plt.plot(times, result2.expect[3], label = 'pump = 1')
plt.xlabel('Time (ns)')
plt.ylabel('$\sigma_{22}$')
legend = plt.legend()
plt.savefig("sigma_22.png")
plt.savefig("sigma_22.pdf")
output_data = np.vstack((times, result1.expect[3]))
file_data_store('sigma_22.dat',output_data.T, numtype = "real")
output_data = np.vstack((times, result2.expect[3]))
file_data_store('sigma_22.dat',output_data.T, numtype = "real")
plt.title('g=120, nth=$10^{-6}$, T1=$10^{-3}$, T2=1')

plt.show()
```

Being basis elements, the sum of the populations should be 1 and phonon number should not go negative but the plots obtain tell a different story:
[phonon number.pdf](https://github.com/qutip/qutip/files/3360514/phonon.number.pdf)
[sigma_00.pdf](https://github.com/qutip/qutip/files/3360515/sigma_00.pdf)
[sigma_11.pdf](https://github.com/qutip/qutip/files/3360516/sigma_11.pdf)
[sigma_22.pdf](https://github.com/qutip/qutip/files/3360517/sigma_22.pdf)

The version I am using is:
```
QuTiP: Quantum Toolbox in Python
Copyright (c) 2011 and later.
A. J. Pitchford, P. D. Nation, R. J. Johansson, A. Grimsmo, and C. Granade

QuTiP Version:      4.4.0.dev0
Numpy Version:      1.16.4
Scipy Version:      1.2.1
Cython Version:     0.29.11
Matplotlib Version: 3.1.0
Python Version:     3.7.3
Number of CPUs:     8
BLAS Info:          INTEL MKL
OPENMP Installed:   False
INTEL MKL Ext:      True
Platform Info:      Windows (AMD64)
```

Any help would mean a lot to me!

**Comment by nwlambert** [comment]
Perhaps someone else can give a more concrete answer, but I guess the driving frequency is too high for the number of time-steps?  Try tweaking the ODE solver with options like the following:

`
options = Options(nsteps=15000, atol=1e-13, rtol=1e-13)     
`

and passing it to mesolve

`result1 = mesolve(Ha, psi0, times, [np.sqrt(T1)*sig01, np.sqrt(T2)*sig02, np.sqrt(gamma*(nth + 1))*a, np.sqrt(gamma*nth)*a.dag()], [a.dag()*a, sig00, sig11, sig22], progress_bar=True, options=options)
`

More practically speaking, for the problem you are solving I guess explicitly including the time dependence is unneccessary, since you have already made a RWA on the drive and the coupling?  

Moving to an interaction picture in w1, w2 and wph, and removing the explicit time-dependence might simplify the problem and might give you equivalent results, and in the process make your code much simpler and more efficient?

In other words just set w1=w2=wph=0, and  put the drive as a time-independent term in the hamiltonian   

`H01 =  g*(sig12*a.dag() + sig21*a) + pump1*(sig01 + sig01.dag())

`

**Comment by XiaoTailong** [comment]
My problem may be more serious. I cannot execute the demo of time-dependent in LZ transition. I guess it is the problem of OS. My platform is MAC OS 10.14 and GCC is 4.2.1. I don't have MS VS 2015. May the environment is not satisfied.

**Comment by xlelephant** [comment]
I also got some wrong result with mesolve, which uses ZVODE integration method. The mesolve raises no warning. But I was able to get the right result with ISODA method. 

The situation was applying a Pi pulse(~50 ns) to the qubit and wait about 1000 ns and then apply another Pi pulse.  The second Pi pulse had no effect for the final state.
I was using [qobj, func] pairs for the time dependent Hamiltonian, where func is the self-defined gate envelope.

**Comment by BoxiLi** [comment]
Hi @xlelephant, it sounds like a different issue you mentioned, could you open a new issue with an example?

**Comment by xlelephant** [comment]
No problem, but to create an example, I need extra time to unwrap the self-defined gate sequence and refactor my test.
